src/plugins/data/public/search/fetch/handle_response.tsx

```diff
@@ -24,44 +24,34 @@ export function handleResponse(
 ) {
   const { rawResponse } = response;
 
-  if (rawResponse.timed_out) {
-    getNotifications().toasts.addWarning({
-      title: i18n.translate('data.search.searchSource.fetch.requestTimedOutNotificationMessage', {
-        defaultMessage: 'Data might be incomplete because your request timed out',
-      }),
-    });
-  }
+  // display warning toast notifications for timeouts and/or shard failures
+  const warnings = extractWarnings(rawResponse);
 
-  if (rawResponse._shards && rawResponse._shards.failed && !disableShardFailureWarning) {
-    const title = i18n.translate('data.search.searchSource.fetch.shardsFailedNotificationMessage', {
-      defaultMessage: '{shardsFailed} of {shardsTotal} shards failed',
-      values: {
-        shardsFailed: rawResponse._shards.failed,
-        shardsTotal: rawResponse._shards.total,
-      },
-    });
-    const description = i18n.translate(
-      'data.search.searchSource.fetch.shardsFailedNotificationDescription',
-      {
-        defaultMessage: 'The data you are seeing might be incomplete or wrong.',
-      }
-    );
+  const timedOut = warnings.find((w) => w.isTimeout === true);
+  if (timedOut) {
+    getNotifications().toasts.addWarning({ title: timedOut.message });
+  }
 
-    const text = toMountPoint(
-      <>
-        {description}
-        <EuiSpacer size="s" />
-        <ShardFailureOpenModalButton
-          request={request.body}
-          response={rawResponse}
-          theme={theme}
-          title={title}
-        />
-      </>,
-      { theme$: theme.theme$ }
-    );
+  const shardFailures = warnings.filter((w) => w.isShardFailure === true);
+  if (shardFailures && !disableShardFailureWarning) {
+    shardFailures.forEach((w) => {
+      const title = w.message;
+      const text = toMountPoint(
```

Comment:
nit: could we construct this in extractWarnings ?

Reply:
I think it would give less value to the consumer to provide a MountPoint instead of the actual messaging